backend/src/test/java/asegroup1/api/services/landregistry/LandRegistryServiceImplTest.java

```diff
@@ -188,45 +190,211 @@ void testIfInvalidPostCodeFetchesNullCoordinates() {
+                    mockRequest.getDouble("top"),
+                    mockRequest.getDouble("right"),
+                    mockRequest.getDouble("bottom"),
+                    mockRequest.getDouble("left")
+            )).thenReturn(postcodesFetchedMock);
+
+            when(landRegistryDataDaoMock.searchForLandRegistryDataInBoundaries(
+                    mockRequest.getDouble("top"),
+                    mockRequest.getDouble("right"),
+                    mockRequest.getDouble("bottom"),
+                    mockRequest.getDouble("left")
+            )).thenReturn(postCodeLocationData);
+
+            landRegistryService = new LandRegistryServiceImpl(landRegistryDataDaoMock);
+
+            assert landRegistryService.getPositionInsideBounds(mockRequest).size() == 1;
+        } catch (Exception e) {
+            e.printStackTrace();
+            assert false;
+        }
+    }
+
+    @Test
+    void testIfPostcodesAreAggregatedCorrectly() {
+        LandRegistryDaoImpl landRegistryDataDaoMock = mock(LandRegistryDaoImpl.class);
+
+        String[] postcodes = {"BN14 7BH", "NW9 9PR", "NN12 8DT", "TW7 4QN", "L22 3YU", "RM17 6LJ", "RG14 7DF", "SE25 5RT"};
+
+
+        List<LandRegistryData> landRegistryDataList = new ArrayList<>();
+
+        for (int i = 0; i < LandRegistryServiceImpl.getAggregationLevels()[1] + 1; i++) {
+            LandRegistryData landRegistryData = new LandRegistryData();
+            landRegistryData.setPostCode(postcodes[i % postcodes.length]);
+
+            landRegistryDataList.add(landRegistryData);
+        }
+
+        when(landRegistryDataDaoMock.searchForLandRegistryDataInBoundaries(0, 0, 0, 0))
+                .thenReturn(landRegistryDataList);
+
+        LandRegistryServiceImpl landRegistryService = new LandRegistryServiceImpl(landRegistryDataDaoMock);
+
+        JSONObject mockRequest = new JSONObject();
+
+        try {
+            mockRequest.put("top", 0);
+            mockRequest.put("right", 0);
+            mockRequest.put("bottom", 0);
+            mockRequest.put("left", 0);
+            for (LandRegistryData positionInsideBound : (List<LandRegistryData>) landRegistryService.getPositionInsideBounds(mockRequest)) {
+                //If addresses are fetched constraint size will be greater than 1
+                assert positionInsideBound.getAllConstraints().size() == 1;
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+
+            assert false;
+        }
+    }
+
+    @Test
+    void testIfHeatMapIsReturnedWhenThresholdIsPassed() {
+        Random random = new Random(RANDOM_SEED);
+
+        LandRegistryDaoImpl landRegistryDataDaoMock = mock(LandRegistryDaoImpl.class);
+
+        String[] postcodes = {"BN14 7BH", "NW9 9PR", "NN12 8DT", "TW7 4QN", "L22 3YU", "RM17 6LJ", "RG14 7DF", "SE25 5RT"};
+
+
+        List<LandRegistryData> landRegistryDataList = new ArrayList<>();
+
+        for (int i = 0; i < LandRegistryServiceImpl.getAggregationLevels()[2] + 1; i++) {
```

Comment:
Instead of adding one to getAggregationLevels you could change the operation to <=